File: extension/utils/masker.js
/**
 * masker.js
 * 탐지된 PII / 인젝션 위치 정보를 받아 텍스트를 마스킹하는 유틸리티
 */

const TYPE_LABELS = {
  // PII
  PERSON_NAME:   '이름',
  EMAIL:         '이메일',
  PHONE:         '전화번호',
  ADDRESS:       '주소',
  ID_NUMBER:     '신분증번호',
  CREDIT_CARD:   '카드번호',
  DATE_OF_BIRTH: '생년월일',
  ORGANIZATION:  '기관명',
  BANK_ACCOUNT:  '계좌번호',
  OTHER_PII:     '개인정보',
  // PII (로컬 encoder 모델, chan/pii TARGET_LABELS)
  PS_NAME:             '이름',
  LC_ADDRESS:           '주소',
  OG_WORKPLACE:         '근무지',
  OG_DEPARTMENT:        '부서명',
  CV_POSITION:          '직위',
  OGG_EDUCATION:        '학력',
  QT_MOBILE:            '휴대폰번호',
  QT_PHONE:             '전화번호',
  QT_RESIDENT_NUMBER:   '주민등록번호',
  QT_ALIEN_NUMBER:      '외국인등록번호',
  QT_DRIVER_NUMBER:     '운전면허번호',
  QT_PLATE_NUMBER:      '차량번호',
  QT_ACCOUNT_NUMBER:    '계좌번호',
  QT_CARD_NUMBER:       '카드번호',
  TMI_EMAIL:            '이메일',
  QT_PASSPORT_NUMBER:   '여권번호',
  QT_AGE:               '나이',
  DT_BIRTH:             '생년월일',
  FD_MAJOR:             '전공',
  // Injection
  INSTRUCTION_OVERRIDE: '명령 재정의',
  ROLE_MANIPULATION:    '역할 조작',
  SYSTEM_PROMPT_LEAK:   '시스템 프롬프트 유출',
  JAILBREAK:            '탈옥 시도',
  HIDDEN_COMMAND:       '숨겨진 명령',
  DATA_EXFILTRATION:    '데이터 유출 시도',
  OTHER_INJECTION:      '인젝션',
};

function getLabel(type) {
  return TYPE_LABELS[type] ?? type;
}

/**
 * items 목록의 위치가 실제 텍스트와 일치하는지 검증하고 보정합니다.
 * LLM이 off-by-one 오류를 낼 수 있으므로 텍스트 매칭으로 보정합니다.
 */
function validateAndFix(text, items) {
  const valid = [];
  for (const item of items) {
    const { start, end, text: itemText } = item;

    // 범위 기본 검증
    if (
      typeof start !== 'number' ||
      typeof end !== 'number' ||
      start < 0 ||
      end > text.length ||
      start >= end
    ) {
      // 텍스트 검색으로 위치 복구 시도
      const idx = text.indexOf(itemText);
      if (idx !== -1) {
        valid.push({ ...item, start: idx, end: idx + itemText.length });
      }
      continue;
    }

    // 위치가 실제 텍스트와 일치하는지 확인
    const slice = text.slice(start, end);
    if (slice === itemText) {
      valid.push(item);
    } else {
      // 약간의 오차 허용 (±5자 범위 탐색)
      let found = false;
      for (let offset = -5; offset <= 5; offset++) {
        const s = start + offset;
        const e = end + offset;
        if (s >= 0 && e <= text.length && text.slice(s, e) === itemText) {
          valid.push({ ...item, start: s, end: e });
          found = true;
          break;
        }
      }
      if (!found) {
        // 전체 검색으로 위치 복구 시도
        const idx = text.indexOf(itemText);
        if (idx !== -1) {
          valid.push({ ...item, start: idx, end: idx + itemText.length });
        }
      }
    }
  }
  return valid;
}

/**
 * 텍스트에서 item 범위들이 겹치지 않도록 병합합니다.
 * 겹치는 범위는 가장 큰 범위로 통합합니다.
 */
function mergeOverlapping(items) {
  if (items.length === 0) return [];
  const sorted = [...items].sort((a, b) => a.start - b.start);
  const merged = [sorted[0]];
  for (let i = 1; i < sorted.length; i++) {
    const last = merged[merged.length - 1];
    const cur = sorted[i];
    if (cur.start < last.end) {
      // 겹침: end를 더 큰 값으로 확장
      last.end = Math.max(last.end, cur.end);
      last.text = items[0].text; // 대표 텍스트 유지
    } else {
      merged.push(cur);
    }
  }
  return merged;
}

/**
 * 텍스트에 마스킹을 적용합니다.
 *
 * @param {string} text           - 원본 텍스트
 * @param {Array}  items          - [{type, text, start, end, ...}]
 * @param {'PII'|'INJECTION'} category
 * @returns {{ maskedText: string, applied: Array }}
 *   applied: 실제 적용된 마스킹 목록 (start/end는 maskedText 기준)
 */
export function applyMasking(text, items, category = 'PII') {
  const validated = validateAndFix(text, items);
  const unique = mergeOverlapping(validated);

  // 뒤에서부터 처리하면 앞쪽 인덱스가 유지됨
  const sorted = [...unique].sort((a, b) => b.start - a.start);

  let result = text;
  const applied = [];

  for (const item of sorted) {
    const label = getLabel(item.type);
    const placeholder = `[${label} 마스킹]`;
    result = result.slice(0, item.start) + placeholder + result.slice(item.end);
    applied.push({ ...item, placeholder, category });
  }

  return { maskedText: result, applied: applied.reverse() };
}

/**
 * 텍스트에서 마스킹된 구간을 강조한 HTML을 생성합니다.
 * (HITL 모달의 "마스킹된 내용" 패널 렌더링용)
 *
 * @param {string} text    - 마스킹된 텍스트
 * @param {'light'|'dark'} theme
 * @returns {string} HTML string
 */
export function buildMaskedHTML(text) {
  // [xxx 마스킹] 패턴을 하이라이트 span으로 변환
  const escaped = escapeHtml(text);
  return escaped.replace(
    /\[([^\]]+) 마스킹\]/g,
    (match, label) => {
      const isPii = !['명령 재정의', '역할 조작', '시스템 프롬프트 유출',
                      '탈옥 시도', '숨겨진 명령', '데이터 유출 시도', '인젝션']
                    .includes(label);
      const cls = isPii ? 'mask-pii' : 'mask-injection';
      return `<span class="mask-tag ${cls}">[${label} 마스킹]</span>`;
    }
  );
}

/**
 * 원본 텍스트에서 PII/인젝션 구간을 강조한 HTML을 생성합니다.
 * (HITL 모달의 "원본 내용" 패널 렌더링용)
 *
 * @param {string} text
 * @param {Array}  piiItems       - PII 탐지 결과
 * @param {Array}  injectionItems - 인젝션 탐지 결과
 * @returns {string} HTML string
 */
export function buildOriginalHTML(text, piiItems, injectionItems) {
  // 모든 span을 합쳐 정렬
  const allItems = [
    ...piiItems.map(i => ({ ...i, category: 'PII' })),
    ...injectionItems.map(i => ({ ...i, category: 'INJECTION' })),
  ].sort((a, b) => a.start - b.start);

  const merged = mergeOverlapping(allItems);
  let html = '';
  let cursor = 0;

  for (const item of merged) {
    if (item.start > cursor) {
      html += escapeHtml(text.slice(cursor, item.start));
    }
    const cls = item.category === 'PII' ? 'highlight-pii' : 'highlight-injection';
    const label = getLabel(item.type);
    html += `<mark class="${cls}" title="${label}">${escapeHtml(text.slice(item.start, item.end))}</mark>`;
    cursor = item.end;
  }

  if (cursor < text.length) {
    html += escapeHtml(text.slice(cursor));
  }

  return html;
}

function escapeHtml(str) {
  return str
    .replace(/&/g, '&amp;')
    .replace(/</g, '&lt;')
    .replace(/>/g, '&gt;')
    .replace(/"/g, '&quot;')
    .replace(/'/g, '&#39;');
}

File: engine/app/core/masker/masker.py
"""
app/core/masker/masker.py
탐지 위치 정보를 받아 텍스트를 마스킹한다.
파이프라인/extension/utils/masker.js 를 파이썬으로 이식:
  - validate_and_fix : 위치 검증·보정(off-by-one, ±5자 탐색, 전체 검색 복구)
  - merge_overlapping: 겹침 병합 (겹치면 더 높은 confidence 유형을 대표로)
  - apply_masking    : 뒤에서부터 치환(앞쪽 인덱스 보존)

PII 는 유형별 라벨([이름 마스킹] 등), 인젝션은 PLAN §4 대로 [인젝션 마스킹] 고정.
"""

from __future__ import annotations

from typing import Any

Item = dict[str, Any]

# PLAN §5 / masker.js TYPE_LABELS (PII 유형 → 한국어 라벨)
TYPE_LABELS: dict[str, str] = {
    "PERSON_NAME": "이름",
    "EMAIL": "이메일",
    "PHONE": "전화번호",
    "ADDRESS": "주소",
    "ID_NUMBER": "신분증번호",
    "CREDIT_CARD": "카드번호",
    "DATE_OF_BIRTH": "생년월일",
    "ORGANIZATION": "기관명",
    "BANK_ACCOUNT": "계좌번호",
    "OTHER_PII": "개인정보",
}

# 인젝션 유형 전체 — 어떤 유형이든 [인젝션 마스킹] 으로 치환 (PLAN §4)
INJECTION_LABEL = "인젝션"

INJECTION_TYPES = {
    "INSTRUCTION_OVERRIDE",
    "ROLE_MANIPULATION",
    "SYSTEM_PROMPT_LEAK",
    "JAILBREAK",
    "HIDDEN_COMMAND",
    "DATA_EXFILTRATION",
    "OTHER_INJECTION",
}


def get_label(item_type: str) -> str:
    if item_type in INJECTION_TYPES:
        return INJECTION_LABEL
    return TYPE_LABELS.get(item_type, "개인정보")


def placeholder_for(item_type: str) -> str:
    return f"[{get_label(item_type)} 마스킹]"


def validate_and_fix(text: str, items: list[Item]) -> list[Item]:
    """위치가 실제 텍스트와 일치하는지 검증·보정 (masker.js validateAndFix 이식).

    항목의 모양 자체가 어긋난 것(dict 이 아니거나 type 이 문자열이 아닌 것)은 여기서
    버린다. 이 함수는 MCP 의 mask_text 를 통해 **외부 입력**(AI 가 채워 보낸 목록)을
    직접 받는 경로에 있어서, 아래 단계들이 item["type"] 을 그대로 인덱싱하면
    KeyError/TypeError 로 500 이 난다.
    """
    valid: list[Item] = []
    for item in items:
        if not isinstance(item, dict) or not isinstance(item.get("type"), str):
            continue
        start = item.get("start")
        end = item.get("end")
        item_text = item.get("text") or ""
        if not isinstance(item_text, str):
            continue

        # 항목에 원문(text)이 없을 수 있다 — MCP 응답은 원문 유출을 막으려고 text 를
        # 빼고 내보낸다(adapters/mcp/tools.py 의 _redact_items). 그 항목을 그대로
        # mask_text 에 되돌려주는 흐름이 정상 경로다.
        #
        # 그때는 내용 대조를 할 수단이 없으니 좌표를 신뢰한다. 이 분기가 없으면 아래
        # 대조가 전부 실패하고 text.find("") 도 -1 이 되어 **항목이 통째로 버려진다** —
        # 마스킹이 하나도 적용되지 않았는데 호출은 성공으로 끝나는 최악의 실패다.
        if not item_text:
            if isinstance(start, int) and isinstance(end, int) and 0 <= start < end <= len(text):
                valid.append(item)
            continue

        if (
            not isinstance(start, int)
            or not isinstance(end, int)
            or start < 0
            or end > len(text)
            or start >= end
        ):
            idx = text.find(item_text) if item_text else -1
            if idx != -1:
                valid.append({**item, "start": idx, "end": idx + len(item_text)})
            continue

        if text[start:end] == item_text:
            valid.append(item)
            continue

        found = False
        for offset in range(-5, 6):
            s = start + offset
            e = end + offset
            if s >= 0 and e <= len(text) and text[s:e] == item_text:
                valid.append({**item, "start": s, "end": e})
                found = True
                break
        if not found:
            idx = text.find(item_text) if item_text else -1
            if idx != -1:
                valid.append({**item, "start": idx, "end": idx + len(item_text)})
    return valid


def merge_overlapping(items: list[Item]) -> list[Item]:
    """겹치는 범위를 병합 (masker.js mergeOverlapping 이식 + 유형 선택 개선).

    겹치면 범위를 합치고, 대표 유형은 confidence 가 높은(동률이면 더 긴) 항목으로
    둔다 — 예: 주민번호(0.98) vs 전화번호가 겹치면 주민번호 유형이 남는다.
    """
    if not items:
        return []
    ordered = sorted(items, key=lambda x: (x["start"], -(x["end"] - x["start"])))
    merged: list[Item] = [dict(ordered[0])]
    for cur in ordered[1:]:
        last = merged[-1]
        if cur["start"] < last["end"]:
            # 겹침 — 더 높은 우선순위(confidence, 길이) 유형을 대표로 채택.
            # confidence 가 None 으로 실려오는 경로가 있다(MCP _redact_items 를 거친
            # 항목, secure_search_files 의 라인 단위 재구성) — or 0.0 으로 받아낸다.
            cur_score = (cur.get("confidence") or 0.0, cur["end"] - cur["start"])
            last_score = (last.get("confidence") or 0.0, last["end"] - last["start"])
            if cur_score > last_score:
                last["type"] = cur["type"]
                last["confidence"] = cur.get("confidence", last.get("confidence", 0.0))
            new_end = max(last["end"], cur["end"])
            last["end"] = new_end
            last["text"] = None  # 병합 후 대표 text 는 apply 시 재산출
        else:
            merged.append(dict(cur))
    return merged


def apply_masking(text: str, items: list[Item]) -> dict[str, Any]:
    """텍스트에 마스킹을 적용 (masker.js applyMasking 이식).

    반환: {"masked_text": str, "applied": [ {type,start,end,placeholder,...} ]}
    (start/end 는 원본 text 기준)
    """
    validated = validate_and_fix(text, items)
    unique = merge_overlapping(validated)

    # 뒤에서부터 처리하면 앞쪽 인덱스가 유지됨
    ordered = sorted(unique, key=lambda x: x["start"], reverse=True)

    result = text
    applied: list[Item] = []
    for item in ordered:
        ph = placeholder_for(item["type"])
        result = result[: item["start"]] + ph + result[item["end"] :]
        applied.append(
            {
                "type": item["type"],
                "start": item["start"],
                "end": item["end"],
                "placeholder": ph,
                "confidence": item.get("confidence"),
                "source": item.get("source"),
            }
        )
    applied.reverse()
    return {"masked_text": result, "applied": applied}
